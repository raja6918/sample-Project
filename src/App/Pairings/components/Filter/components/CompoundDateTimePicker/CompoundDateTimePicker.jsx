import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Grid } from '@material-ui/core';
import DateFnsUtils from '@date-io/date-fns';
import { DateTimePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import { DateRange } from '@material-ui/icons';
import moment from 'moment';

const PickersWrapper = styled.div`
  .picker-field {
    margin-top: 5px;
    & div > input {
      cursor: pointer;
    }
  }
  .picker-field2 {
    margin-top: 16px;
    & div > input {
      cursor: pointer;
    }
  }
  & .errorSpan {
    color: #d10000 !important;
    margin-top: 3px;
    font-size: 0.75rem;
  }
`;

class CompoundDateTimePicker extends Component {
  state = {
    startTime: this.props.value.startTime,
    endTime: this.props.value.endTime,
    error: false,
  };

  componentDidMount() {
    if (this.props.value.noChange) {
      this.sendDateTime();
    }
  }

  isDateTimeRangeValid = () => {
    const { startTime, endTime } = this.state;
    const { returnDateFormat } = this.props;
    if (startTime && endTime) {
      const startDateTimeF = moment(startTime, returnDateFormat);
      const endDateTimeF = moment(endTime, returnDateFormat);
      return startDateTimeF.isSameOrBefore(endDateTimeF);
    }
    return true;
  };

  sendDateTime = () => {
    const { startTime, endTime, error } = this.state;
    const { returnDateFormat, onChange } = this.props;
    const startDateTimeF = moment(startTime).format(returnDateFormat);
    const endDateTimeF = moment(endTime).format(returnDateFormat);
    let value = {};

    if (startTime && !endTime) value = { startTime: startDateTimeF };
    else if (endTime && !startTime) value = { endTime: endDateTimeF };
    else if (endTime && startTime)
      value = { startTime: startDateTimeF, endTime: endDateTimeF };

    onChange(value, null, error);
  };

  handleChange = (event, field) => {
    this.setState({ [field]: event }, () =>
      this.setState({ error: !this.isDateTimeRangeValid() }, this.sendDateTime)
    );
  };

  render() {
    const { t, displayDateFormat, containerStyle } = this.props;
    const { startTime, endTime, error } = this.state;

    return (
      <Grid container style={containerStyle}>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <PickersWrapper>
            <Grid item xs={12} sm={12}>
              <DateTimePicker
                name="startTime"
                ampm={false}
                className="picker-field"
                label={t('FILTER.pane.labels.between')}
                maxDate={endTime || null}
                format={displayDateFormat}
                placeholder={displayDateFormat.toUpperCase()}
                value={startTime}
                onChange={event => this.handleChange(event, 'startTime')}
                InputProps={{
                  endAdornment: (
                    <DateRange color="primary" style={{ cursor: 'pointer' }} />
                  ),
                }}
              />
            </Grid>
            <Grid item xs={12} sm={12}>
              <DateTimePicker
                className="picker-field2"
                name="endTime"
                ampm={false}
                label={t('FILTER.pane.labels.and')}
                minDate={startTime || null}
                format={displayDateFormat}
                placeholder={displayDateFormat.toUpperCase()}
                value={endTime}
                onChange={event => this.handleChange(event, 'endTime')}
                InputProps={{
                  endAdornment: (
                    <DateRange color="primary" style={{ cursor: 'pointer' }} />
                  ),
                }}
              />
            </Grid>
            {error && (
              <span className="errorSpan">
                {t('ERRORS.FILTER.dateTimeRangeIncorrect')}
              </span>
            )}
          </PickersWrapper>
        </MuiPickersUtilsProvider>
      </Grid>
    );
  }
}

CompoundDateTimePicker.propTypes = {
  t: PropTypes.func.isRequired,
  value: PropTypes.shape({
    startTime: PropTypes.string.isRequired,
    endTime: PropTypes.string.isRequired,
    noChange: PropTypes.bool,
  }),
  onChange: PropTypes.func.isRequired,
  returnDateFormat: PropTypes.string,
  displayDateFormat: PropTypes.string,
  containerStyle: PropTypes.shape({}),
};

CompoundDateTimePicker.defaultProps = {
  returnDateFormat: 'YYYY-MM-DDTHH:mm',
  displayDateFormat: 'yy/MM/dd HH:mm',
  value: { startTime: null, endTime: null, noChange: true },
  containerStyle: { width: '160px' },
};

export default CompoundDateTimePicker;
