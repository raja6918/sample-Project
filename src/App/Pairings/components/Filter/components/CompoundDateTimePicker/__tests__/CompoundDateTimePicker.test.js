import React from 'react';
import { shallow } from 'enzyme';
import CompoundDateTimePicker from '../CompoundDateTimePicker';

describe('Test CompoundDateTimePicker Component', () => {
  const minProps = {
    t: jest.fn(),
    value: {
      startTime: '2020-12-06T06:02',
      endTime: '2020-12-06T06:02',
    },
    onChange: jest.fn(),
    returnDateFormat: 'YYYY-MM-DDTHH:mm',
    displayDateFormat: 'yy/MM/dd HH:mm',
    containerStyle: { width: '160px' },
  };

  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<CompoundDateTimePicker {...minProps} />);
  });

  test('CompoundDateTimePicker Component renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('isDateTimeRangeValid returns true by when dates are not set', () => {
    wrapper.instance().setState({
      startTime: null,
      endTime: null,
    });
    const result = wrapper.instance().isDateTimeRangeValid();
    expect(result).toBeTruthy();
  });

  test('isDateTimeRangeValid returns false when the start date is equal to the end date', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:55',
      endTime: '20/12/06 06:55',
    });
    const result = wrapper.instance().isDateTimeRangeValid();
    expect(result).toBeTruthy();
  });

  test('isDateTimeRangeValid returns false when the start date is greater than the end date', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:56',
      endTime: '20/12/06 06:55',
    });
    const result = wrapper.instance().isDateTimeRangeValid();
    expect(result).toBeFalsy();
  });

  test('isDateTimeRangeValid returns true when the start date is lesser than the end date', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:54',
      endTime: '20/12/06 06:55',
    });
    const result = wrapper.instance().isDateTimeRangeValid();
    expect(result).toBeTruthy();
  });

  test('handleChange sets the error when the date values are not valid', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:54',
      endTime: '20/12/06 06:54',
    });
    const event = '20/12/06 06:53';
    const field = 'endTime';
    wrapper.instance().handleChange(event, field);
    expect(wrapper.state('error')).toBeTruthy();
  });

  test('handleChange unsets the error when the date values are valid', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:54',
      endTime: '20/12/06 06:53',
    });
    const event = '20/12/06 06:55';
    const field = 'endTime';
    wrapper.instance().handleChange(event, field);
    expect(wrapper.state('error')).toBeFalsy();
  });

  test('sendDateTime works correctly', () => {
    wrapper.instance().setState({
      startTime: '20/12/06 06:54',
      endTime: '20/12/06 06:53',
    });
    wrapper.instance().sendDateTime();

    expect(minProps.onChange).toBeCalledWith(
      expect.objectContaining({
        startTime: expect.any(String),
        endTime: expect.any(String),
      }),
      null,
      expect.any(Boolean)
    );
  });
});
