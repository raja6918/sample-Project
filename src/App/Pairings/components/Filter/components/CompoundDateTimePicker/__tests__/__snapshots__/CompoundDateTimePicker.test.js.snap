// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CompoundDateTimePicker Component CompoundDateTimePicker Component renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  style={
    Object {
      "width": "160px",
    }
  }
>
  <MuiPickersUtilsProvider
    utils={[Function]}
  >
    <CompoundDateTimePicker__PickersWrapper>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={12}
        xs={12}
      >
        <PickerWithState
          InputProps={
            Object {
              "endAdornment": <Memo
                color="primary"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              />,
            }
          }
          allowKeyboardControl={true}
          ampm={false}
          className="picker-field"
          format="yy/MM/dd HH:mm"
          invalidDateMessage="Invalid Date Format"
          maxDate="2020-12-06T06:02"
          maxDateMessage="Date should not be after maximal date"
          minDate={1900-01-01T00:00:00.000Z}
          minDateMessage="Date should not be before minimal date"
          name="startTime"
          onChange={[Function]}
          openTo="date"
          orientation="portrait"
          placeholder="YY/MM/DD HH:MM"
          showTabs={true}
          value="2020-12-06T06:02"
          views={
            Array [
              "year",
              "date",
              "hours",
              "minutes",
            ]
          }
          wider={true}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        sm={12}
        xs={12}
      >
        <PickerWithState
          InputProps={
            Object {
              "endAdornment": <Memo
                color="primary"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
              />,
            }
          }
          allowKeyboardControl={true}
          ampm={false}
          className="picker-field2"
          format="yy/MM/dd HH:mm"
          invalidDateMessage="Invalid Date Format"
          maxDate={2100-01-01T00:00:00.000Z}
          maxDateMessage="Date should not be after maximal date"
          minDate="2020-12-06T06:02"
          minDateMessage="Date should not be before minimal date"
          name="endTime"
          onChange={[Function]}
          openTo="date"
          orientation="portrait"
          placeholder="YY/MM/DD HH:MM"
          showTabs={true}
          value="2020-12-06T06:02"
          views={
            Array [
              "year",
              "date",
              "hours",
              "minutes",
            ]
          }
          wider={true}
        />
      </WithStyles(ForwardRef(Grid))>
    </CompoundDateTimePicker__PickersWrapper>
  </MuiPickersUtilsProvider>
</WithStyles(ForwardRef(Grid))>
`;
